internal_task_handlers.go

```diff
@@ -724,45 +728,82 @@ func newActivityTaskHandler(activities []activity,
 }
 
 type cadenceInvoker struct {
-	identity      string
-	service       m.TChanWorkflowService
-	taskToken     []byte
-	cancelHandler func()
-	retryPolicy   backoff.RetryPolicy
+	identity               string
+	service                m.TChanWorkflowService
+	taskToken              []byte
+	cancelHandler          func()
+	retryPolicy            backoff.RetryPolicy
+	heartBeatTimeoutInSec  int32 // The heart beat interval configured for this activity.
+	lastHeartBeatTimestamp time.Time
 }
 
 func (i *cadenceInvoker) Heartbeat(details []byte) error {
+	nonTransientError := false
+	if !i.isTimeToReportHeartbeat() {
+		return nil
+	}
+
 	err := recordActivityHeartbeat(i.service, i.identity, i.taskToken, details, i.retryPolicy)
 
 	switch err.(type) {
 	case *CanceledError:
 		// We are asked to cancel. inform the activity about cancellation through context.
 		// We are asked to cancel. inform the activity about cancellation through context.
 		i.cancelHandler()
+		nonTransientError = true
 
 	case *s.EntityNotExistsError:
 		// We will pass these through as cancellation for now but something we can change
 		// later when we have setter on cancel handler.
 		i.cancelHandler()
+		nonTransientError = true
+	}
+
+	if err == nil || nonTransientError {
+		i.lastHeartBeatTimestamp = time.Now()
 	}
 
 	// We don't want to bubble temporary errors to the user.
 	// This error won't be return to user check RecordActivityHeartbeat().
 	return err
 }
 
+func (i *cadenceInvoker) isTimeToReportHeartbeat() bool {
+	// If we don't have any heartbeat timeout configured.
+	lastHeartbeatTS := i.lastHeartBeatTimestamp
+	if i.heartBeatTimeoutInSec <= 0 {
+		if lastHeartbeatTS.IsZero() || time.Now().After(lastHeartbeatTS.Add(defaultHeartBeatInterval)) {
+			return true
+		}
+		return false
+	}
+
+	// If we have heartbeat timeout configured.
+	// The heartbeat time indicates that on any two successive pings the distance should not be more than the
+	// the configured timeout.
+	// We want to suppress reporting HB too frequently to server, until we get auto heartbeating.
+	// we will suppress reporting HB if it is with-in a certain window(60% of HB timeout)
+	duration := time.Duration(0.6*float32(i.heartBeatTimeoutInSec)) * time.Second
```

Comment:
this won't work. assume we have activity with heartbeat timeout configured to 100s, and the activity starts and immediately heartbeat, lets say it heartbeat at time 0, then at time 59s, it heart beat again, by the proposed rule, this heartbeat will be suppressed, then the activity continues its work, and at time 150s, it heartbeat again, which is supposed to be 91s after its previous heartbeat and should not timeout, but since we suppressed the previous one, it will timeout.